config/src/runtime_upgrades.rs

```diff
@@ -0,0 +1,278 @@
+use std::sync::{MutexGuard, PoisonError};
+
+pub use serde::{Deserialize, Serialize};
+use thiserror::*;
+
+type Result<T, E = ReloadFailure> = std::result::Result<T, E>;
+
+/// Error which occurs when reloading a configuration fails.
+#[derive(Clone, Copy, Debug, Error)]
+pub enum ReloadFailure {
+    /// Locking the Mutex within [`ReloadHandle`] failed.
+    #[error("Cannot acquire Mutex Lock. Poisoned Mutex.")]
+    PoisonedMutex,
+
+    /// If the reload handle wasn't properly initialised (using
+    /// [`ReloadHandle::set`]), there's nothing to reaload with.
+    #[error("Cannot reload an uninitialised handle.")]
+    NotInitialised,
+
+    /// Error not specified by the implementors of the [`Reloadable`]
+    /// traits. Use as last resort.
+    #[error("Unspecified reload failure.")]
+    Other,
+}
+
+impl<T>
+    From<
+        PoisonError<
+            MutexGuard<'_, std::option::Option<Box<(dyn Reloadable<T> + Send + Sync + 'static)>>>,
+        >,
+    > for ReloadFailure
+{
+    fn from(
+        _: PoisonError<
+            MutexGuard<'_, std::option::Option<Box<(dyn Reloadable<T> + Send + Sync + 'static)>>>,
+        >,
+    ) -> Self {
+        Self::PoisonedMutex
+    }
+}
+
+/// Implement this trait if the field must be mutably borrowed for `reload`.
+pub trait Reloadable<T> {
+    /// Reload `self` using provided `item`.
+    ///
+    /// # Errors
+    /// Fails with an appropriate variant of
+    /// [`ReloadFailure`]. [`ReloadFailure::Other`] can be used as a
+    /// **temporary** placeholder.
+    fn reload(&mut self, item: T) -> Result<()>;
+}
+
+/// Implement this trait if the object can be immutably borrowed and `reload`-ed.
+pub trait InnerMutablyReloadable<T> {
+    /// Reload `self` using provided `item`.
+    ///
+    /// # Errors
+    /// Fails with an appropriate variant of [`ReloadFailure`].
+    /// [`ReloadFailure::Other`] can be used as a **temporary** placeholder.
+    fn reload(&self, item: T) -> Result<()>;
+}
+
+impl<T> std::fmt::Debug for dyn Reloadable<T> {
```

Comment:
alternative could be to write a blanket implementation like:
`impl<T, K: Reload<T>> std::fmt::Debug for K`